In [1]:
import matplotlib.pyplot as plt
import numpy as np
import sympy as sp
from scipy import linalg as la
sp.init_printing()
import warnings
warnings.filterwarnings("ignore")

## Goals and Commands

#### Goals
1. Dot Producs
2. Norm
3. Gram-Schmidt Process
4. QR-factorization

#### Commands
1. norm()
2. GramSchmidt()
3. linalg.qr()


#### Dot Product
There are few ways to do dot product in Jupyter:

In [3]:
v=np.array([1,2,3])
u=np.array([-1,2,5])

In [4]:
# matrix multiplication
u@v

np.int64(18)

In [5]:
# using inner product command (inner product is more general name for dot product)
np.inner(u,v)

np.int64(18)

In [6]:
# Product with SymPy
a=sp.Matrix(u)
b=sp.Matrix(v)
a.T*b

[18]

To find norm is easy too:

In [7]:
np.linalg.norm(u)

np.float64(5.477225575051661)

In [8]:
la.norm(u)

np.float64(5.477225575051661)

In [11]:
a.norm()

√30

### Example

Let $A=\begin{bmatrix} 1 & 1 & 2 \\ 2 & 5 & 0 \\ 4 & 7 & 4 \end{bmatrix}$. Show that Row(A) is orthogonal to Nul(A)

#### Solution

In [12]:
A=sp.Matrix([[1,1,2],[2,5,0],[4,7,4]])
row=A.rowspace()
nul=A.nullspace()
row,nul

⎛                         ⎡⎡-10/3⎤⎤⎞
⎜                         ⎢⎢     ⎥⎥⎟
⎜[[1  1  2], [0  3  -4]], ⎢⎢ 4/3 ⎥⎥⎟
⎜                         ⎢⎢     ⎥⎥⎟
⎝                         ⎣⎣  1  ⎦⎦⎠

Note both row and nul are lists containing vectors, so we will have to use indecies. To check orthogonality we just need to find all pairwise dot products. Since, row contains row vectors, we don't need to use transpose:

In [13]:
row[0]*nul[0]

[0]

In [14]:
row[1]*nul[0]

[0]

Both dot products are zero. Hence, both spaces are orthogonal to each other.
#### End of solution

Next, lets explore null space commands in more details:

In [15]:
A=np.array([[1,1,2],[5,5,10],[4,4,8]])
N=la.null_space(A)
N

array([[ 0.91287093,  0.        ],
       [-0.18257419, -0.89442719],
       [-0.36514837,  0.4472136 ]])

In [16]:
v1=N[:,0]
v2=N[:,1]

In [17]:
v1@v2

np.float64(-1.1102230246251565e-16)

In [18]:
la.norm(v1)

1.0

In [19]:
la.norm(v2)

1.0

As you can see, la.null_space produces a matrix whose columns form orthonormal basis for Null space. Note that v1@v2=1.11022302462516 e−16. While this not zero, it is very very close to zero. In fact, it isn't a zero only due to a rounding error.

The SymPy version (nullspace() ) doesn't provide orthonormal basis.

### Gram-Schmidt Process

SymPy includes are direct way to do Gram-Schmidt process:

#### Example

Orthogonalize column space for $A=\begin{bmatrix} 1 & 1 & 0 & 3\\ 2 & 5 & -1 & 3\\ 0 & 1 & 2 & 0\end{bmatrix}$

#### Solution

In [22]:
A=sp.Matrix([[1,1,0,3],[2,5,-1,3],[0,1,2,0]])
Col=A.columnspace()
sp.GramSchmidt(Col)


⎡⎡1⎤  ⎡-6/5⎤  ⎡ 1  ⎤⎤
⎢⎢ ⎥  ⎢    ⎥  ⎢    ⎥⎥
⎢⎢2⎥, ⎢3/5 ⎥, ⎢-1/2⎥⎥
⎢⎢ ⎥  ⎢    ⎥  ⎢    ⎥⎥
⎣⎣0⎦  ⎣ 1  ⎦  ⎣3/2 ⎦⎦

In [23]:
# Gram-Schmidt with normalization
sp.GramSchmidt(Col, orthonormal=True)

⎡        ⎡-3⋅√70 ⎤  ⎡ √14 ⎤⎤
⎢⎡ √5 ⎤  ⎢───────⎥  ⎢ ─── ⎥⎥
⎢⎢ ── ⎥  ⎢  35   ⎥  ⎢  7  ⎥⎥
⎢⎢ 5  ⎥  ⎢       ⎥  ⎢     ⎥⎥
⎢⎢    ⎥  ⎢ 3⋅√70 ⎥  ⎢-√14 ⎥⎥
⎢⎢2⋅√5⎥, ⎢ ───── ⎥, ⎢─────⎥⎥
⎢⎢────⎥  ⎢  70   ⎥  ⎢ 14  ⎥⎥
⎢⎢ 5  ⎥  ⎢       ⎥  ⎢     ⎥⎥
⎢⎢    ⎥  ⎢  √70  ⎥  ⎢3⋅√14⎥⎥
⎢⎣ 0  ⎦  ⎢  ───  ⎥  ⎢─────⎥⎥
⎣        ⎣  14   ⎦  ⎣ 14  ⎦⎦

#### End of Solution

### QR-factorization

This is an important type of marix factorization for many reasons. One of them is that in A=QR, the columns of Q are orthonormal basis of Col(A). This, in one go, provides a way to find basis for column space and to orthogonalize a set of vectors using NumPy or SciPy. (Note while sympy can do QR factorization as well, usually you have to manually round numbers as it will always try to be precise and use roots. Therefore, usually NumPy is a prefered way).

QR factorization will be even more important later when we discuss least-square approximation and PCA.

#### Example

Find orthonormal basis of $A=\begin{bmatrix} 1 & 2 & 0 \\ 1 & 5 & 1 \\ 0 & -1 & 2 \\ 3& 3 & 0\end{bmatrix}$  using QR-factorization

#### Solution

In [34]:
A=np.array([[1,2,0],[1,5,1],[0,-1,2], [3,3,0]])
Q, R = np.linalg.qr(A)

In [35]:
Q

array([[-0.30151134, -0.1375409 ,  0.06625475],
       [-0.30151134, -0.89401584, -0.25706844],
       [-0.        ,  0.25215831, -0.96201901],
       [-0.90453403,  0.34385225,  0.06360456]])

In [36]:
R

array([[-3.31662479, -4.82418151, -0.30151134],
       [ 0.        , -3.96576257, -0.38969921],
       [ 0.        ,  0.        , -2.18110647]])

In [37]:
# Let's check the norms of columns of Q
la.norm(Q[:,0]), la.norm(Q[:,1]), la.norm(Q[:,2])

(1.0, 1.0, 1.0)

In [38]:
# Let's check orthogonality
Q[:,0]@Q[:,1], Q[:,0]@Q[:,2],Q[:,1]@Q[:,2]

(8.326672684688674e-17, 0.0, 5.551115123125783e-17)

#### End of Solution

Finally here is the SymPy way:

In [39]:
sp.matrices.MatrixBase.QRdecomposition(sp.Matrix(A))

⎛⎡ √11    6⋅√1903    -25⋅√142379 ⎤                         ⎞
⎜⎢ ───    ───────    ────────────⎥                         ⎟
⎜⎢ 11      1903         142379   ⎥  ⎡     16⋅√11    √11   ⎤⎟
⎜⎢                               ⎥  ⎢√11  ──────    ───   ⎥⎟
⎜⎢ √11    39⋅√1903    97⋅√142379 ⎥  ⎢       11      11    ⎥⎟
⎜⎢ ───    ────────    ────────── ⎥  ⎢                     ⎥⎟
⎜⎢ 11       1903        142379   ⎥  ⎢     √1903   17⋅√1903⎥⎟
⎜⎢                               ⎥, ⎢ 0   ─────   ────────⎥⎟
⎜⎢        -√1903     363⋅√142379 ⎥  ⎢      11       1903  ⎥⎟
⎜⎢  0     ───────    ─────────── ⎥  ⎢                     ⎥⎟
⎜⎢          173        142379    ⎥  ⎢             √142379 ⎥⎟
⎜⎢                               ⎥  ⎢ 0     0     ─────── ⎥⎟
⎜⎢3⋅√11  -15⋅√1903   -24⋅√142379 ⎥  ⎣               173   ⎦⎟
⎜⎢─────  ──────────  ────────────⎥                         ⎟
⎝⎣ 11       1903        142379   ⎦                         ⎠